### PR TITLE
Helm charts for the different Landscaper deployers

### LINKED ISSUE #77: Helm Charts automatic version update
**How to categorize this issue?**

/area usability
/kind enhancement
/priority normal

**What would you like to be added**:

Currently the app version of the helm charts that is used as the default image tag is only updated for the oci helm artifact during its release.

The master helm chart which is mostly used by the community still refers to `0.1.0`.

To be up to date, the helm charts versions should be automatically updated to the latest release version.

### LINKED ISSUE #104: Make Deployers installable
**How to categorize this issue?**

/area delivery
/kind task
/priority normal

**What would you like to be added**:

Currently only the landscaper itself and the container deployer have helm charts so that they can be installed on their own.
In order to run the landscaper and its deployers in production it is needed that the deployers can be deployed on their.

In addition, it should be possible to deploy for example the container deployer with a blueprint (with the helm deployer).
Therefore it would be possible to fully manage deployers with just the helm deployer.

Therefore we would need to perform the following tasks:
- [x] Create Helm Chart for all Deployers
  - Helm
  - Manifest
  - Mock (no priority)
- [x] Add the Helm charts as build artifacts. See the container deployer as an example https://github.com/gardener/landscaper/blob/master/.ci/publish-helm-charts
- [x] Create Blueprints for the Deployers (although we need to discuss what we choose as the initial deployer (helm or manifest))
- [x] Create Integration tests that install the deployers and runs 1/2 simple integration tests
  - With that it would make sense to have dedicated integration tests that only work with raw DeployItems and no landscaper involved.

### PR DESCRIPTION
**Pull request by MrBatschner**
**How to categorize this PR?**

/area ops-productivity
/kind enhancement
/priority normal

**What this PR does / why we need it**:
This PR adds Helm charts for the individual dedicated deployers of Landscaper together with their integration in the publishing pipeline.

**Which issue(s) this PR fixes**:
Fixes #104
Also fixes https://github.com/gardener/landscaper/issues/77

**Special notes for your reviewer**:
none atm

**Release note**:

```other operator
Landscaper's Deployers can now be deployed from their very own Helm charts.
```

### CODE REVIEW

File: .landscaper/container-deployer/blueprint.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        apiVersion: helm.deployer.landscaper.gardener.cloud/v1alpha1
+        kind: ProviderConfiguration
+        updateStrategy: update
+        name: {{ .imports.releaseName }}
+        namespace: {{ .imports.releaseNamespace }}
+        chart:
+          {{ $resource := getResource .cd "name" "container-deployer-chart" }}
+          ref: {{ $resource.access.imageReference }}
+        values:
+          {{ toYaml .imports.values | indent 4 }}
```
Comment (schrodit) [inline .landscaper/container-deployer/blueprint.yaml:37]: the controller image reference `image.repository` should be read from the component descriptor

File: .landscaper/container-deployer/example/imports.yaml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    targetSelector:
+      - annotations:
+        - key:
+          operator:
+          value:
+
+    replicaCount: 1
+
+    image:
+      repository: eu.gcr.io/gardener-project/landscaper/container-deployer-controller
```
Comment (schrodit) [inline .landscaper/container-deployer/example/imports.yaml:31]: should not be set by any input data as it is read from the component descriptor

File: charts/helm-deployer/templates/clusterrole.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  name: {{ include "deployer.fullname" . }}
+  labels:
+    {{- include "deployer.labels" . | nindent 4 }}
+rules:
+- apiGroups:
+  - landscaper.gardener.cloud
+  resources:
+  - deployitems
+  - deployitems/status
+  - targets
```
Comment (schrodit) [inline charts/helm-deployer/templates/clusterrole.yaml:18]: the deployers do need update permissions for targets. read should be enough

File: hack/fix-helm-chart-version.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
Comment (schrodit) [inline hack/fix-helm-chart-version.sh:1]: would rather name the file `update-helm-chart-version` or `set-helm-chart-version`.
Fix does more indicate some error

File: test/integration/helmcharts/chartstests.go [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+			const deployerName = "container-deployer"
+
+			target, err := utils.CreateInternalKubernetesTarget(ctx, f.Client, state.Namespace, deployerName, f.RestConfig)
+			utils.ExpectNoError(err)
+			utils.ExpectNoError(state.Create(ctx, f.Client, target))
+
+			di := forgeDeployItem(chartDir, valuesFile, deployerName, target)
+			utils.ExpectNoError(state.Create(ctx, f.Client, di))
+			utils.ExpectNoError(utils.WaitForHelmDeployItemToSucceed(ctx, f.TestLog(), f.Client, kutil.ObjectKey(di.Name, state.Namespace), 2*time.Minute))
+			utils.ExpectNoError(utils.WaitForDeploymentToBeReady(ctx, f.TestLog(), f.Client, kutil.ObjectKey(deployerName, state.Namespace), 2*time.Minute))
```
Comment (schrodit) [inline test/integration/helmcharts/chartstests.go:71]: I think we should also test the deletion afterwards.

File: test/integration/helmcharts/chartstests.go [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+		})
+	})
+}
+
+func deployHelmDeployer(ctx context.Context, state *envtest.State, f *framework.Framework, dumper *framework.Dumper) {
```
Comment (schrodit) [inline test/integration/helmcharts/chartstests.go:140]: the helm deployer is installed by creating a deployitem of type helm with the helmdeployer as base64 encoded chart and then call the helm specific go code to deploy the item?

File: test/integration/helmcharts/chartstests.go [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+			state, cleanup, err := f.NewState(ctx)
+			utils.ExpectNoError(err)
+			dumper.AddNamespaces(state.Namespace)
+			defer func() {
+				Expect(cleanup(ctx)).ToNot(HaveOccurred())
+			}()
+
+			By("Setting up the Helm deployer first")
+			deployHelmDeployer(ctx, state, f, dumper)
```
Comment (schrodit) [inline test/integration/helmcharts/chartstests.go:54]: I think the helm deployer should be also gracefully deleted afterwards.
Currently the deletion is done by the state which will just remove the deploy items but the helm deployer is leaked?

File: test/integration/helmcharts/chartstests.go [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+
+	exports := make(map[string]interface{})
+
+	err = h.ApplyFiles(ctx, files, exports)
+	Expect(err).NotTo(HaveOccurred())
+
+	utils.ExpectNoError(utils.WaitForDeploymentToBeReady(ctx, f.TestLog(), f.Client, kutil.ObjectKey(helmDeployerName, state.Namespace), 2*time.Minute))
+}
+
+func forgeDeployItem(chartDir string, valuesFile string, name string, target *lsv1alpha1.Target) *lsv1alpha1.DeployItem {
```
Comment (schrodit) [inline test/integration/helmcharts/chartstests.go:171]:
```suggestion
func forgeHelmDeployItem(chartDir string, valuesFile string, name string, target *lsv1alpha1.Target) *lsv1alpha1.DeployItem {
```

File: test/utils/charts.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+	"k8s.io/apimachinery/pkg/util/wait"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	lsv1alpha1 "github.com/gardener/landscaper/apis/core/v1alpha1"
+	"github.com/gardener/landscaper/pkg/utils/simplelogger"
+)
+
+// WaitForHelmDeployItemToSucceed waits for a DeployItem to succeed
+func WaitForHelmDeployItemToSucceed(ctx context.Context, logger simplelogger.Logger, kubeClient client.Client, objKey types.NamespacedName, timeout time.Duration) error {
```
Comment (schrodit) [inline test/utils/charts.go:29]: This function could be used for any kind of deploy items not only helm charts?

**Review by schrodit — CHANGES_REQUESTED** [review]
looks good overall.

Currently the landscaper is deployed with all deployers for the integration tests.
We need to ensure in the helm deployer integration that the deploy items are not picked up by the central deployer

File: test/integration/helmcharts/chartstests.go (hunk 6 above)
Comment (MrBatschner) [reply test/integration/helmcharts/chartstests.go:140]: Chicken-Egg problem. We need the Helm deployer to deploy Helm charts so how do we get the Helm deployer into the cluster?
However, this was based on the assumption that we do not have a landscaper-controller with integrated deployers in the test cluster (as it was the case for my local setup). If we we have the landscaper-controller with enabled internal helm deployer available, then this code is indeed not needed.

File: test/utils/charts.go (hunk 9 above)
Comment (MrBatschner) [reply test/utils/charts.go:29]: Yep, I will rename it and move it to `test/utils/landscaper.go`.

File: hack/fix-helm-chart-version.sh (hunk 4 above)
Comment (MrBatschner) [reply hack/fix-helm-chart-version.sh]: `update-helm-chart-version.sh` it is.

File: charts/helm-deployer/templates/clusterrole.yaml (hunk 3 above)
Comment (MrBatschner) [reply charts/helm-deployer/templates/clusterrole.yaml]: Fixed for all deployers, `targets` are now `get, list, watch` only.

File: test/integration/helmcharts/chartstests.go (hunk 6 above)
Comment (schrodit) [reply test/integration/helmcharts/chartstests.go:140]: currently it is like that. And I think we also planned afterwards to keep a lightwight version. 
So lets keep it simple and expect a running deployer

Comment (MrBatschner) [reply test/integration/helmcharts/chartstests.go]: Fixed, the code that deploys the initial Helm-Deployer is removed and the landscaper-controller's internal Helm-deployer is now responsible for picking up the DeployItems.
The Deployers that get deployed during the test will have an active TargetSelector so that they will not pick up any DeployItem (unless they have a matching annotation of course).

File: test/integration/helmcharts/chartstests.go (hunk 7 above)
Comment (MrBatschner) [reply test/integration/helmcharts/chartstests.go]: Obsolete with the manual deployment of the Helm deployer removed.

File: test/integration/helmcharts/chartstests.go (hunk 5 above)
Comment (MrBatschner) [reply test/integration/helmcharts/chartstests.go]: Done.

File: .landscaper/container-deployer/blueprint.yaml (hunk 1 above)
Comment (MrBatschner) [reply .landscaper/container-deployer/blueprint.yaml]: Fixed, image ref will now be taken from component descriptor and merged with the other values for the Helm charts that come from an import.

File: .landscaper/container-deployer/example/imports.yaml (hunk 2 above)
Comment (MrBatschner) [reply .landscaper/container-deployer/example/imports.yaml]: Fixed.

**Comment by MrBatschner** [comment]
Incorporated all suggestions and improved integration test for the deployer Helm charts.
@schrodit Please have another look.

Integration test fails currently for the Mock deployer as we do not have a publish step for its image so far and therefore, the Mock deployer cannot be deployed to the test-machinery. The publish step for the Mock deployer gets introduced with this PR. All other steps are succeeding just fine.

File: charts/landscaper/Chart.yaml [hunk 10]
```diff
@@ -19,9 +19,9 @@ type: application
 # This is the chart version. This version number should be incremented each time you make changes
 # to the chart and its templates, including the app version.
 # Versions are expected to follow Semantic Versioning (https://semver.org/)
-version: 0.1.0
+version: v0.5.3
```
Comment (schrodit) [inline charts/landscaper/Chart.yaml:22]: The chart version can stay at 0.1.0 only the appVersion should be increased

File: charts/manifest-deployer/Chart.yaml [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+#
+# Library charts provide useful utilities or functions for the chart developer. They're included as
+# a dependency of application charts to inject those utilities and functions into the rendering
+# pipeline. Library charts do not define any templates and therefore cannot be deployed.
+type: application
+
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: v0.5.3
```
Comment (schrodit) [inline charts/manifest-deployer/Chart.yaml:22]: see comment above

File: test/utils/utils.go [hunk 12]
```diff
@@ -5,9 +5,11 @@
 package utils
 
 import (
+	"github.com/onsi/ginkgo"
 	"github.com/onsi/gomega"
 )
 
 func ExpectNoError(err error, extra ...interface{}) {
+	defer ginkgo.GinkgoRecover()
```
Comment (schrodit) [inline test/utils/utils.go:13]: why is this needed?
isn't this already done by the fail handler registration in the test.

**Review by schrodit — CHANGES_REQUESTED** [review]
just some minor comments.
And the mock deployer test is failing can you have a look at that

File: charts/landscaper/Chart.yaml (hunk 10 above)
Comment (MrBatschner) [reply charts/landscaper/Chart.yaml]: Changed for all Helm charts and also in the `update-helm-chart-version.sh` script.

File: charts/manifest-deployer/Chart.yaml (hunk 11 above)
Comment (MrBatschner) [reply charts/manifest-deployer/Chart.yaml]: Fixed.

**Comment by MrBatschner** [comment]
> And the mock deployer test is failing can you have a look at that

As I mentioned, it will fail as long as this PR did not get merged to master. We currently do not have a CI/CD step for publishing the mock-deployer's container image to GCR so that specific test is bound to fail (the mock-deployers Pod will fail with ErrImagePull). 
Since this PR does introduce the publishing step for the mock-deployer, once it is merged to master, all future tests will succeed.

File: hack/create-helm-chart.sh [hunk 13]
```diff
@@ -35,7 +35,8 @@ cli.py config attribute --cfg-type container_registry --cfg-name gcr-readwrite -
 echo "> Creating helm chart ${CHART_NAME}:${EFFECTIVE_VERSION} from $CHART_PATH"
 
 # update version and appVersion
-sed -i -e "s/0.1.0/${EFFECTIVE_VERSION}/" ${PROJECT_ROOT}/${CHART_PATH}/Chart.yaml
+sed -i -e "s/^version:.*/version: ${EFFECTIVE_VERSION}/" ${PROJECT_ROOT}/${CHART_PATH}/Chart.yaml
```
Comment (schrodit) [inline hack/create-helm-chart.sh:38]: This does not need to be updated
```suggestion
```

**Review by schrodit — APPROVED** [review]
/lgtm

File: hack/create-helm-chart.sh (hunk 13 above)
Comment (MrBatschner) [reply hack/create-helm-chart.sh]: Forgot that one :unamused: - fixed.
